### ISSUE TITLE
[experimental] Change search panel to a single horizontal textbox

### ISSUE THREAD

**Issue by yurishkuro** [uber/jaeger-ui#20]
It could be an alternative way of displaying search params for advanced users. The idea is that you just type things into the textbox and some intelli-sense helper provides auto-completion. E.g.:

```
service:hystrix tag:debug-id=myID time:last-30-min
```

**Comment by tiffon** [comment]
A couple of options around implementation:

- *[Code Mirror](http://codemirror.net/demo/complete.html)* with autocomplete - Stays true to the text field input, not sure how well it handles very large lists of complete options, might have to limit to some sub-set (e.g. first 50 entries)
- *[VisualSearch.js](https://documentcloud.github.io/visualsearch/)* - Deviates a bit from the text field input, also very dated, maybe makes sense to implement from scratch, but interesting prior art
- *[Select2](https://select2.github.io/examples.html)* - A rock-solid dropdown component. Can be used, in series, to emulate the type of text field input you've described. Can be a bit nuanced to implement, but I've had some success with this approach in the past.
- From scratch with `<div contenteditable>` 

Probably the best implementation of this feature, that I've seen, is the search input in the [Smart Search](https://www.jetbrains.com/youtrack/features/issue_tracking.html) feature of YouTrack.

![youtrack_smart_search](https://cloud.githubusercontent.com/assets/2304337/26182908/5a17f5ee-3b49-11e7-8f32-7f8218ea0f3d.png)

**Comment by yurishkuro** [comment]
Yes, the YouTrack implementation looks very similar to what I had in mind. As long as the generic auto-completion component is implemented, the actual query syntax can be fairly straightforward to implement as a simple state machine.

**Comment by saminzadeh** [comment]
+ 1 I think this is a great idea.

Follow up question: Should this just be UI sugar, or is it better to have a query DSL on the backend that can interpret these queries?

**Comment by yurishkuro** [comment]
I think it should be just the UI feature. Because auto-complete is critical for such feature, it means the UI will already have a structured data model representing the query, not just a free form text, therefore there's no reason why the backend should be exposed to that.

BTW, to add a bit more, I think it would be useful to display already parsed / recognized tokens similar to how GitHub does it for labels:

![image](https://cloud.githubusercontent.com/assets/3523016/26209210/2a3b2adc-3bba-11e7-89e9-80e6471454d6.png)

**Comment by tiffon** [comment]
As a reference point, I took a look at the search interface in Facebook's ad management UI.

FB has taken a more structured approach. This seems appropriate for complex search / filter functionality. As it stands, the Jaeger UI search is straight-forward, but if there are plans to make it more expressive then a similarly structured approach might be worth considering. Features like the following are simple in and of themselves, but they add up quickly if a number of them are included:

- Search on multiple services / operations at once via `AND` or `OR`
- `AND` two search queries together
- etc.

Another aspect of the structured approach is discoverability for new users. This would only matter if the search functionality is made more expressive.

Some screenshots:

Top level attributes
![fb_00_attributes](https://cloud.githubusercontent.com/assets/2304337/26516942/0659c192-425d-11e7-9014-83a62d37c9b2.png)

Operators
![fb_01_operator](https://cloud.githubusercontent.com/assets/2304337/26516946/065b1b82-425d-11e7-9144-34e272a003ba.png)

Freeform text value input
![fb_02_clause_freeform_text](https://cloud.githubusercontent.com/assets/2304337/26516943/065a556c-425d-11e7-8f03-0289cfaba3c5.png)

Fully applied search clause
![fb_03_clause_is_applied](https://cloud.githubusercontent.com/assets/2304337/26516947/065cf740-425d-11e7-9366-0e4fca3ac5e1.png)

Multi-select search value input
![fb_05_clause_value_multi-select](https://cloud.githubusercontent.com/assets/2304337/26516945/065b2370-425d-11e7-9866-2b04e03de227.png)

Multiple values selected
![fb_06_clause_multiple_values](https://cloud.githubusercontent.com/assets/2304337/26516944/065a8550-425d-11e7-823c-e6e4fd2223fd.png)

Predefined filters
![fb_07_predefined_filters](https://cloud.githubusercontent.com/assets/2304337/26516953/066b167c-425d-11e7-9fcc-cf3156aebe88.png)

Date input
![fb_08_date_input](https://cloud.githubusercontent.com/assets/2304337/26516949/0667e402-425d-11e7-9110-5e39bf5cabc9.png)

Date applied
![fb_09_date_applied](https://cloud.githubusercontent.com/assets/2304337/26516948/06667b1c-425d-11e7-9c44-77bf6eb79dfd.png)

Date range selection
![fb_10_date_range_selection](https://cloud.githubusercontent.com/assets/2304337/26516950/0668c840-425d-11e7-8647-1f22d81b8abd.png)

Hints in the dropdown
![fb_12_misc_dropdown_hints](https://cloud.githubusercontent.com/assets/2304337/26516951/0669fe36-425d-11e7-8467-37616337a3ac.png)

**Comment by yurishkuro** [comment]
Overall I like it, but one thing I don't like is that it seems the filter cannot be entered (or removed) without using the mouse. The way I imagined this working in a textbox is similar to these screenshots, except the Add `[+]` happens automatically as you start typing something, and removing a filter can be done with a backspace (hit once - highlight the filter on the left, hit twice - delete it).

**Comment by tiffon** [comment]
IMO a UX along the lines of what you've described would be ideal. An
example of the approach is the recipient entry in gmail, which also allows
editing on `enter` in addition to deletion on `delete` or `backspace`. It
should be possible to achieve it with a structured approach. It can be a
bit tricky, but I've done it in the past with a filter implementation.

**Comment by tiffon** [comment]
I vote we kill this ticket. With the width of the left column being adjustable and the full text showing on mouseover, I agree with @jcorbin, this ticket is unnecessary. 

@jcorbin @yurishkuro Care to weigh in?

**Comment by Dieterbe** [comment]
I'd be curious to hear from @yurishkuro why advanced users would prefer this interface over the current interface (or potentially an improved version of the current interface, but with the same concept of a form of multiple fields).  I suspect advanced users would prefer it if they don't like hitting tab to go from one form field to the next, especially skipping over form fields they don't need; and generally it's less distracting I think if they can do everything on the same horizontal line, instead of making your eyes jump all over a form.
So the main differentiator seems to me how we switch between /navigate across different filters/settings, but the editing experience of individual filters/settings can probably be made very similar between the two different UI approaches (not sure though)

Having provided a similar query input approach to users in the past (http://vimeo.github.io/graph-explorer) the feedback i constantly got is that new users wanted a very obvious way to be gently guided into all the searching/filtering possibilities. that said, my query input was nowhere near as helpful with autocompletions and widgets as yours is, but still the warning holds.
this of course raises the question about whether this will be an alternative/additional UI for power-users only (which doesn't need to be as beginner friendly), or the one and only interface (in which case we have to try really hard to make it intuitive and not put anything in the way hindering users who are unfamiliar with a lot of the concepts in trying to issue their first queries)

I generally like the concept though.
I like especially the first 2 pre-filled filters, but then it gets a tad confusing.
why does it first say "Search" but sometimes becomes "Tag, Min Duration ..." and back to "search" based on where you click? it gives the impression you can type an unstructured text seach query but that doesn't seem possible. (note that the current UI makes it very obvious what all the options are - probably because i'm still a fairly new user - i appreciate that a lot) 
maybe it can set up the default state in such a way that it's more obvious what the possibilities are?
once i go past the first 2 widgets i get a bit confused about what I can and/or should enter.
